docs/source/notes/source_of_recompilation.md

```diff
@@ -0,0 +1,176 @@
+ c = torch.rand(s, device='xla') # c can be of any shape between [0, 5] which causes more recompilations!
+ d = c + 1
+```
+
+So this one is actually hard to solve without PyTorch frontend’s help. What do we need?
+
+In short, we need a Tensor world!
+
+For example,
+
+* `tensor.size()` should return a Tensor so that it can be a Tensor with dynamic shape and kept in the graph without early evaluation.
+* Tensor accessor, e.g. for 2D tensor, `tensor[0][0]` now returns a value but this need to return a tensor as well.
+* Implicitly this means all operators currently taking int/float/double as input need a Tensor overload as well. THIS IS A BIG ASK as it can easily explode our operator set.
+    * It’s easier if we can make scalar to Tensor conversion really cheap so that we can only care about the Tensor overload.
+    * In practice not all ops takes scalars from previous computation, so we’ve been adding Tensor variants by ad-hoc requests.
+    * This is also a common ask from tracing base approaches I think.
+
+Okay now that we assume every op in PyTorch has a Tensor verison we need, are we done?
+
+## #3. From control flow
+
+No! We actually only solved the problem without data dependent control flow...
+
+See the example below:
+
+```
+if x[0][0] == 3:
+  bla
+else:
+  blabla
+```
+
+Even if `x[0][0]` was a Tensor, we need to execute/materialize its value for python interpreter to proceed. And different branch choices in multiple control flows combined means we have a lot of graph to compile as well!
+
+For now we just have no way to fix this. To fix it we need to lower the control flow from python to graph! Without too much thinking in implementation we can do this in two ways:
+
+* ask users to explicitly use a control flow op instead of python if/else/while/for. This is currently supported as [customized API in torch_xla](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_builder.py#L563-L574) but not widely adopted in users’ code. (python users are used to if/else/for and it’s hard to switch them to a uglier API unless there’s a huge perf win).
+* parse python source. code to get the control flow statement automatically. This is like Torchscript and somehow merge the torchscripted graph into the lazily trace graph properly (including shape info etc). I haven’t thought through the steps of how to implement this indeed :P
+
+But either solution above requires non-trivial amount of effort, either on user side or on the framework side. That’s why we currently just take the hit of early evaluation & multiple compilations as a short term solution given the bandwidth we have.
+
+Okay so now we assume that also have control flow lowered in the graph automagically, are we gold?
+
+YES! Now you have your whole computation represented in a graph of Tensor operations, including control flow so that compilers can now consume and do their smart tricks! But tbh at this point your program is no longer very PyTorch-y.
```

Comment:
Ftr - TF itself takes this approach and there are a lot of expressivity gaps, even before you further reduce to HLO, which had even tighter constraints. It falls into the category of things that people do when they have no choice, but it is painful to watch users try to navigate the land mines :/